compiler/rustc_mir/src/interpret/intern.rs

```diff
@@ -356,7 +356,7 @@ where
                 // an allocation, which we should avoid. When that happens,
                 // dedicated error variants should be introduced instead.
                 assert!(
-                    !error.kind.allocates(),
+                    !error.kind().allocates(),
```

Comment:
Doesn't this stop making any sense now?
The point of this `allocates` thing was to make sure we don't allocate an error and then "catch" it (for perf reasons we wanted to avoid those allocations). But with this change we now allocate for all errors, so (a) it seems silly to also separately `Box` some error kinds *again*, and (b) this check here makes no sense since we always allocate. So shouldn't we get rid of the extra nested `Box` and the `allocates` stuff?

(Btw I'd appreciate a Cc for changes like this that affect the Miri/CTFE engine.)

Reply:
Unboxing of `UndefinedBehaviorInfo::InvalidUninitBytes` content would be one thing to consider, but generally this check makes as much sense now as it made before. Heap allocation inside & outside cannot be easily merged together.

Reply:
> this check makes as much sense now as it made before

If that means it didn't make any sense before, that's okay and we should still remove it. ;)

But I feel like the difference between 0 and 1 allocation is much bigger than between 1 and 2, so I am not sure this check is worth keeping. OTOH, maybe the expensive bit we should protect against isn't the allocation itself but the string formatting in the `format!`-using errors. In that case we should probably rename `allocates` to `pre_formatted` or so (and unbox `InvalidUninitBytes` as you said).